### PR TITLE
Thread local in frame access

### PR DESCRIPTION
**Pull request by aardvark179**
Refactors access to the special variables `$~` and `$_` to share common code, correctly change values in ThreadLocalObjects rather than creating new ThreadLocalObjects, and resolve performance issues with the underlying ThreadLocal maps in the common case where ThreadLocalObjects are only ever accessed from a single thread.

### CODE REVIEW

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+            super(fd, fs, defaultValue);
+            this.depth = depth;
+        }
+
+        @Override
+        public ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame) {
+            final MaterializedFrame frame = RubyArguments.getDeclarationFrame(callerFrame, depth);
+            return getThreadLocalObjectFromFrame(context, frame, fs, defaultValue, true);
+        }
+    }
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java]: I don't think there is any need to separate these code paths.
SimpleStrategy is just ComplexStrategy with depth=0 and
`RubyArguments.getDeclarationFrame(frame, depth);` supports depth=0 and optimizes just as good.

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            final Frame nextFrame = RubyArguments.getDeclarationFrame(frame);
+            if (nextFrame != null) {
+                frame = nextFrame;
+            } else {
+                return frame;
+            }
+        }
+    }
+
+    private static FrameSlot getVaraibleSlot(Frame frame, String variableName) {
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java]: typo

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalObject.java [hunk 3]
```diff
@@ -25,9 +29,50 @@ public static ThreadLocalObject wrap(RubyContext context, Object value) {
...
+        } else {
+            synchronized (this) {
+                if (otherThreadValues != null) {
+                    return otherThreadValues;
+                } else {
+                    return otherThreadValues = new ThreadLocal<>();
+                }
+            }
+        }
+    }
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalObject.java]: Do you know if there are tests/specs exercising this code?

File: truffleruby/src/main/java/org/truffleruby/parser/BodyTranslator.java [hunk 4]
```diff
@@ -1685,10 +1683,12 @@ public RubyNode visitGlobalAsgnNode(GlobalAsgnParseNode node) {
...
-                rhs = new CheckMatchVariableTypeNode(rhs);
-                rhs = WrapInThreadLocalNodeGen.create(rhs);
-                rhs.unsafeSetSourceSection(sourceSection);
-                environment.declareVarInMethodScope("$~");
+                environment.declareVarInMethodScope(name);
+                // The right hand side must be built using the resolved variable.
+                break;
+            case "$_":
+                // The right hand side must be built using the resolved variable.
+                environment.declareVar(name);
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/parser/BodyTranslator.java]: Should these two use the same `declareX` method?

Comment (aardvark179) [inline truffleruby/src/main/java/org/truffleruby/parser/BodyTranslator.java]: Not sure why the original code declared them differently. I've changed them both to declare the variable in the method scope so we are at least consistent.

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+import com.oracle.truffle.api.frame.FrameSlotKind;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.object.DynamicObject;
+
+public abstract class ThreadLocalInFrameNode extends RubyBaseNode {
+
+    protected final String variableName;
+    protected final int limit;
+
+    public ThreadLocalInFrameNode(String variableName, int limit) {
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java]: I think the limit should be just be in a `getLimit()` method returning `getContext().getOptions().FRAME_VARIABLE_ACCESS_LIMIT`, so it doesn't need to be passed and it's immediately clear what option affects it.

File: truffleruby/src/main/java/org/truffleruby/core/regexp/RegexpNodes.java [hunk 6]
```diff
@@ -749,30 +618,11 @@ private static Frame getMatchDataFrameSearchingDeclarations(Frame topFrame, bool
...
-            RubyContext context,
-            Frame topFrame,
-            boolean add) {
-
-        final Frame frame = getMatchDataFrameSearchingDeclarations(topFrame, add);
-        return getThreadLocalObject(context, frame, add);
-    }
-
     @TruffleBoundary
     public static ThreadLocalObject getMatchDataThreadLocalSearchingStack(RubyContext context) {
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/core/regexp/RegexpNodes.java]: The other `getMatchData*` methods can probably also go away if `rb_backref_get` gets refactored into something simpler and based on get some (n) caller frame + ThreadLocalInFrameNode. But good to leave this for another PR.

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/GetFromThreadLocalNode.java [hunk 7]
```diff
@@ -31,14 +34,13 @@ public Object isDefined(VirtualFrame frame) {
...
-        if (RubyGuards.isThreadLocal(threadLocalObject)) {
+        if (isThreadLocalProfile.profile(RubyGuards.isThreadLocal(threadLocalObject))) {
             return getThreadLocalValue((ThreadLocalObject) threadLocalObject);
         }
 
         return threadLocalObject;
     }
 
-    @TruffleBoundary
     private Object getThreadLocalValue(ThreadLocalObject threadLocalObject) {
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/GetFromThreadLocalNode.java]: `ThreadLocal.get` is too big for partial evaluation, so it should remain a boundary.

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+            return matches(Layouts.PROC.getDeclarationFrame(block));
+        }
+
+        public abstract ThreadLocalObject getThreadLocal(RubyContext context, MaterializedFrame callerFrame);
+
+        public ThreadLocalObject getThreadLocalBlock(RubyContext context, DynamicObject block) {
+            return getThreadLocal(context, Layouts.PROC.getDeclarationFrame(block).materialize());
+        }
+
+        public static SetThreadLocalInFrameStrategy ofBlock(RubyContext context, DynamicObject block, String varName) {
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalInFrameNode.java]: This seems to never be used anymore, so I think the whole strategy abstraction can be removed.

File: truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalObject.java [hunk 9]
```diff
@@ -25,9 +29,50 @@ public static ThreadLocalObject wrap(RubyContext context, Object value) {
...
+            fallbackSet(value);
+        }
+    }
+
+    @TruffleBoundary
+    private void fallbackSet(Object value) {
+        getOtherThreadValues().set(value);
     }
 
-    @Override
```
Comment (eregon) [inline truffleruby/src/main/java/org/truffleruby/language/threadlocal/ThreadLocalObject.java]: This `@Override` should remain.
